node/src/components/consensus/highway_core/highway/vertex.rs

```diff
@@ -277,3 +278,64 @@ impl<C: Context> Endorsements<C> {
         self.endorsers.iter().map(|(v, _)| *v)
     }
 }
+
+/// A ping sent by a validator to signal that it is online but has not created new units in a
+/// while.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+#[serde(bound(
+    serialize = "C::Hash: Serialize",
+    deserialize = "C::Hash: Deserialize<'de>",
+))]
+pub(crate) struct Ping<C>
+where
+    C: Context,
+{
+    creator: ValidatorIndex,
+    timestamp: Timestamp,
+    signature: C::Signature,
+}
+
+impl<C: Context> Ping<C> {
+    /// Creates a new signed ping.
+    pub(crate) fn new(
+        creator: ValidatorIndex,
+        timestamp: Timestamp,
+        sk: &C::ValidatorSecret,
+        rng: &mut NodeRng,
+    ) -> Self {
+        Ping {
+            creator,
+            timestamp,
+            signature: sk.sign(&Self::hash(creator, timestamp), rng),
+        }
+    }
+
+    /// The creator who signals that it is online.
+    pub(crate) fn creator(&self) -> ValidatorIndex {
+        self.creator
+    }
+
+    /// The timestamp when the ping was created.
+    pub(crate) fn timestamp(&self) -> Timestamp {
+        self.timestamp
+    }
+
+    /// Validates the ping and returns an error if it is not signed by the creator.
+    pub(crate) fn validate(
+        &self,
+        validators: &Validators<C::ValidatorId>,
+    ) -> Result<(), VertexError> {
+        let v_id = validators.id(self.creator).ok_or(PingError::Creator)?;
+        let hash = Self::hash(self.creator, self.timestamp);
+        if !C::verify_signature(&hash, v_id, &self.signature) {
+            return Err(PingError::Signature.into());
+        }
+        Ok(())
+    }
```

Comment:
I don't think validity of a model should be in the model itself.

Reply:
[It's the same for evidence](https://github.com/CasperLabs/casper-node/blob/b2e4b0485325896447c7d3e60477ba83ecd7b260/node/src/components/consensus/highway_core/evidence.rs#L81).

Where do you think it should be? Should all vertex type validation functions be `Highway` methods? Isn't it cleaner to have them in the module belonging to their type?

Reply:
Maybe it _looks_ cleaner but it doesn't seem right. I'm OK with verifying signature in the model but testing whether the creator is a validator – this should be part of the `Highway`. It's up to a specific controller how to interpret validity. 

Anyway, not a blocker and I certainly won't ask you to change it in this PR. Just something I wanted to share.

Reply:
OK, I'll look into it; maybe I'll create a follow-up PR.